SCAutolib/controller.py

```diff
@@ -221,136 +253,132 @@ def setup_user(self, user_dict: dict, force: bool = False):
-                                    local=False)
+                                    password=user_dict["passwd"])
             if force:
-                if new_user.cert and new_user.cert.exists():
-                    self.ipa_ca.revoke_cert(new_user.cert)
                 new_user.delete_user()
-            user_dict["card_dir"].mkdir(exist_ok=True)
-
             new_user.add_user()
+        self.users.append(new_user)
+        dump_to_json(new_user)
+        return new_user
 
-        if user_dict["card_type"] == "virtual":
-            hsm_conf = file.SoftHSM2Conf(
-                new_user.card_dir.joinpath("sofhtsm2.conf"),
-                new_user.card_dir)
+    def setup_card(self, card_dict: dict, force: bool = False):
+        """
+        Create card object. Card object should contain its root CA cert as it
+        represents general card (i.e. including physical read-only cards).
+
+        :param card_dict: Dictionary containing card attributes
+        :type card_dict: dict
+        :param force: If its true and card directory exists it will be removed
+        :type force: bool
+        """
+        card_dir: Path = Path("/root/cards", card_dict["name"])
+        card_dir.mkdir(parents=True, exist_ok=True)
+
+        if force and card_dir.exists():
+            rmtree(card_dir)
+
+        if card_dict["card_type"] == "physical":
+            new_card = card.PhysicalCard(card_dict, card_dir=card_dir)
+        elif card_dict["card_type"] == "virtual":
+            hsm_conf = file.SoftHSM2Conf(card_dir.joinpath("sofhtsm2.conf"),
+                                         card_dir=card_dir)
             hsm_conf.create()
             hsm_conf.save()
 
-            new_user.card = card.VirtualCard(new_user,
-                                             softhsm2_conf=hsm_conf.path)
+            new_card = card.VirtualCard(card_dict, softhsm2_conf=hsm_conf.path,
+                                        card_dir=card_dir)
+
+            # card needs to know some details of its user, so we add user as
+            # card attribute
+            for card_user in self.users:
+                if card_user.username == card_dict["cardholder"]:
+                    new_card.user = card_user
+
+            if new_card.user.user_type == "local":
+                cnf_path = new_card.card_dir.joinpath(f"{new_card.cardholder}.cnf")
+                cnf = file.OpensslCnf(filepath=cnf_path, conf_type="user",
+                                      replace=[new_card.cardholder, new_card.CN])
+                cnf.create()
+                cnf.save()
+
+                new_card.cnf = cnf.path
+
+                if force:
+                    if new_card.cert and new_card.cert.exists():
+                        self.local_ca.revoke_cert(new_card.cert)
+            else:  # IPA user
+                if force:
+                    if new_card.cert and new_card.cert.exists():
+                        self.ipa_ca.revoke_cert(new_card.cert)
```

Comment:
it would be nice to reduce the size of this single method and move this code to separate function just for better separation of concerns and potential testability